File: +ndi/+daq/+reader/+mfdaq/cedspike2.m
classdef cedspike2 < ndi.daq.reader.mfdaq
% NDI_DAQSYSTEM_MFDAQ_CEDSPIKE2 - Device driver for CED Spike2
%
% This class reads data from CED Spike2 .SMR or .SON file formats.
%
% It depends on sigTOOL by Malcolm Lidierth (http://sigtool.sourceforge.net).
%
% sigTOOL is also included in the https://github.com/VH-Lab/vhlab-thirdparty-matlab bundle and
% can be installed with instructions at http://code.vhlab.org.
%
	properties
		

	end % properties

	methods
		function obj = cedspike2(varargin)
			% NDI_DAQSYSTEM_MFDAQ_CEDSPIKE2 - Create a new NDI_DEVICE_MFDAQ_CEDSPIKE2 object
			%
			%  D = NDI_DAQSYSTEM_MFDAQ_CEDSPIKE2(NAME,THEFILENAVIGATOR)
			%
			%  Creates a new NDI_DAQSYSTEM_MFDAQ_CEDSPIKE2 object with name NAME and associated
			%  filenavigator THEFILENAVIGATOR.
			%
			obj = obj@ndi.daq.reader.mfdaq(varargin{:})
		end

		function channels = getchannelsepoch(ndi_daqreader_mfdaq_cedspike2_obj, epochfiles)
			% GETCHANNELS - List the channels that are available on this device
			%
			%  CHANNELS = GETCHANNELS(THEDEV, EPOCHFILES)
			%
			%  Returns the channel list of acquired channels in this session
			%
			% CHANNELS is a structure list of all channels with fields:
			% -------------------------------------------------------
			% 'name'             | The name of the channel (e.g., 'ai1')
			% 'type'             | The type of data stored in the channel
			%                    |    (e.g., 'analogin', 'digitalin', 'image', 'timestamp')
			% 'time_channel'     | The channel number that has the time information for that channel
			%

				channels = vlt.data.emptystruct('name','type','time_channel');

				multifunctiondaq_channel_types = ndi.daq.system.mfdaq.mfdaq_channeltypes();

				% open SMR files, and examine the headers for all channels present
				%   for any new channel that hasn't been identified before,
				%   add it to the list
				filename = ndi_daqreader_mfdaq_cedspike2_obj.cedspike2filelist2smrfile(epochfiles);

				header = read_CED_SOMSMR_header(filename);

				if isempty(header.channelinfo),
					channels = vlt.data.emptystruct('name','type','time_channel');
				end;

				for k=1:length(header.channelinfo),
					newchannel.type = ndi_daqreader_mfdaq_cedspike2_obj.cedspike2headertype2mfdaqchanneltype(header.channelinfo(k).kind);
					newchannel.name = [ ndi.daq.system.mfdaq.mfdaq_prefix(newchannel.type) int2str(header.channelinfo(k).number) ];
					newchannel.time_channel = header.channelinfo(k).number;
					channels(end+1) = newchannel;

					if strcmp(newchannel.type,'analog_in'), % add the time channel
						newtimechannel.type = 'time';
						newtimechannel.name = [ ndi.daq.system.mfdaq.mfdaq_prefix(newtimechannel.type) int2str(header.channelinfo(k).number) ];
						newtimechannel.time_channel = header.channelinfo(k).number;
						channels(end+1) = newtimechannel;
					end;
				end
		end % getchannels()

		function [b,msg] = verifyepochprobemap(ndi_daqreader_mfdaq_cedspike2_obj, epochprobemap, epochfiles)
			% VERIFYEPOCHPROBEMAP - Verifies that an EPOCHPROBEMAP is compatible with a given device and the data on disk
			%
			%   B = VERIFYEPOCHPROBEMAP(NDI_DAQSYSTEM_MFDAQ_CEDSPIKE2_OBJ, EPOCHPROBEMAP, EPOCHFILES)
			%
			% Examines the ndi.epoch.epochprobemap_daqsystem EPOCHPROBEMAP and determines if it is valid for the given device
			% epoch NUMBER.
			%
			% For the abstract class ndi.daq.system, EPOCHPROBEMAP is always valid as long as
			% EPOCHPROBEMAP is an ndi.epoch.epochprobemap_daqsystem object.
			%
			% See also: ndi.daq.system, ndi.epoch.epochprobemap_daqsystem
				b = 1;
				msg = '';
				% UPDATE NEEDED
		end

		function data = readchannels_epochsamples(ndi_daqreader_mfdaq_cedspike2_obj, channeltype, channel, epochfiles, s0, s1)
			%  FUNCTION READ_CHANNELS - read the data based on specified channels
			%
			%  DATA = READ_CHANNELS(MYDEV, CHANNELTYPE, CHANNEL, EPOCHFILES, S0, S1)
			%
                        %  CHANNELTYPE is the type of channel to read (cell array of strings, one per
                        %     channel, or single string for all channels)
			%
			%  CHANNEL is a vector of the channel numbers to read, beginning from 1
			%
			%  EPOCHFILES is the cell array of full path filenames for this epoch
			%
			%  DATA is the channel data (each column contains data from an indvidual channel) 
			%

				if ~iscell(channeltype),
					channeltype = repmat({channeltype},numel(channel),1);
				end;
				uniquechannel = unique(channeltype);
				if numel(uniquechannel)~=1,
					error(['Only one type of channel may be read per function call at present.']);
				end

				filename = ndi_daqreader_mfdaq_cedspike2_obj.cedspike2filelist2smrfile(epochfiles);
				sr = ndi_daqreader_mfdaq_cedspike2_obj.samplerate(epochfiles, channeltype, channel);
				sr_unique = unique(sr); % get all sample rates
				if numel(sr_unique)~=1,
					error(['Do not know how to handle different sampling rates across channels.']);
				end;


				sr = sr_unique;

				t0 = (s0-1)/sr;
				t1 = (s1-1)/sr;

				if isinf(t0) | isinf(t1),
					t0_orig = t0;
					t1_orig = t1;
					t0t1_here = ndi_daqreader_mfdaq_cedspike2_obj.t0_t1(epochfiles);
					if isinf(t0_orig),
						if t0_orig<0,
							t0 = t0t1_here{1}(1);
						elseif t0_orig>0,
							t0 = t0t1_here{1}(2);
						end;
					end;
					if isinf(t1_orig),
						if t1_orig<0,
							t1 = t0t1_here{1}(1);
						elseif t1_orig>0
							t2 = t0t1_here{1}(2);
						end;
					end;
				end;

				for i=1:length(channel), % can only read 1 channel at a time
					if strcmpi(channeltype,'time'),
						[dummy,dummy,dummy,dummy,data(:,i)] = read_CED_SOMSMR_datafile(filename,'',channel(i),t0,t1);
					else,
						[data(:,i)] = read_CED_SOMSMR_datafile(filename,'',channel(i),t0,t1);
					end
				end

		end % readchannels_epochsamples

		function t0t1 = t0_t1(ndi_daqreader_mfdaq_cedspike2_obj, epochfiles)
			% EPOCHCLOCK - return the t0_t1 (beginning and end) epoch times for an epoch
			%
			% T0T1 = T0_T1(NDI_DAQSYSTEM_MFDAQ_CEDSPIKE2_OBJ, EPOCHFILES)
			%
			% Return the beginning (t0) and end (t1) times of the EPOCHFILES that define this
			% epoch in the same units as the ndi.time.clocktype objects returned by EPOCHCLOCK.
			%
			%
			% See also: ndi.time.clocktype, EPOCHCLOCK
			%
				filename = ndi_daqreader_mfdaq_cedspike2_obj.cedspike2filelist2smrfile(epochfiles);
				header = read_CED_SOMSMR_header(filename);

				t0 = 0;  % developer note: the time of the first sample in spike2 is not 0 but 0 + 1/4 * sample interval; might be more correct to use this
				t1 = header.fileinfo.dTimeBase * header.fileinfo.maxFTime * header.fileinfo.usPerTime;
				t0t1 = {[t0 t1]};
		end % t0t1

		function [timestamps,data] = readevents_epochsamples_native(ndi_daqreader_mfdaq_cedspike2_obj, channeltype, channel, epochfiles, t0, t1)
			%  FUNCTION READEVENTS_EPOCHSAMPLES_NATIVE - read events or markers of specified channels for a specified epoch
			%
			%  DATA = READEVENTS_EPOCHSAMPLES(MYDEV, CHANNELTYPE, CHANNEL, EPOCHFILES, T0, T1)
			%
			%  CHANNELTYPE is the type of channel to read
			%  ('event','marker', etc)
			%
			%  CHANNEL is a vector with the identity of the channel(s) to be read.
			%
			%  EPOCH is the set of epoch files
			%
			%  DATA is a two-column vector; the first column has the time of the event. The second
			%  column indicates the marker code. In the case of 'events', this is just 1. If more than one channel
			%  is requested, DATA is returned as a cell array, one entry per channel.
			%
				%disp('reading here')
                if ~iscell(channeltype),
                    channeltype = repmat({channeltype},numel(channel),1);
                end;
				timestamps = {};
				data = {};
				filename = ndi_daqreader_mfdaq_cedspike2_obj.cedspike2filelist2smrfile(epochfiles);
				for i=1:numel(channel),
					[data{i},dummy,dummy,dummy,timestamps{i}]= ndr.format.ced.read_SOMSMR_datafile(filename, ...
						'',channel(i),t0,t1);
                    if strcmp(channeltype{i},'event'),
                        data{i} = ones(size(data{i}));
                    end;
                    if strcmp(channeltype{i},'marker'),
                        data{i} = sum( double(data{i}) .* repmat([2^0 2^8 2^16 2^24],size(data{i},1),1),2);
                    end;
                    if strcmp(channeltype{i},'text'),
                        data{i} = vlt.data.matrow2cell(data{i})';
                    end;
				end
				if numel(channel)==1,
					timestamps = timestamps{1};
					data = data{1};
				end;
		end % readevents_epochsamples_native()

		function sr = samplerate(ndi_daqreader_mfdaq_cedspike2_obj, epochfiles, channeltype, channel)
			% SAMPLERATE - GET THE SAMPLE RATE FOR SPECIFIC EPOCH AND CHANNEL
			%
			% SR = SAMPLERATE(DEV, EPOCHFILES, CHANNELTYPE, CHANNEL)
			%
			% SR is the list of sample rate from specified channels

				filename = ndi_daqreader_mfdaq_cedspike2_obj.cedspike2filelist2smrfile(epochfiles);

				sr = [];
				for i=1:numel(channel),
					sr(i) = 1/read_CED_SOMSMR_sampleinterval(filename,[],channel(i));
				end

		end % samplerate()

                function [datatype,p,datasize] = underlying_datatype(ndi_daqreader_mfdaq_cedspike2_obj, epochfiles, channeltype, channel)
			% UNDERLYING_DATATYPE - get the underlying data type for a channel in an epoch
			%
			% [DATATYPE,P,DATASIZE] = UNDERLYING_DATATYPE(DEV, EPOCHFILES, CHANNELTYPE, CHANNEL)
			%
			% Return the underlying datatype for the requested channel.
			%
			% DATATYPE is a type that is suitable for passing to FREAD or FWRITE
			%  (e.g., 'float64', 'uint16', etc. See help fread.)
			%
			% P is a polynomial that converts between the double data that is returned by
			% READCHANNEL. RETURNED_DATA = (RAW_DATA+P(1))*P(2)+(RAW_DATA+P(1))*P(3) ...
			%
			% DATASIZE is the sample size in bits.
			%                                       
			% CHANNELTYPE must be a string. It is assumed that
			% that CHANNELTYPE applies to every entry of CHANNEL.
			%                          


				switch(channeltype),            
					case {'analog_in','analog_out'},
           				filename = ndi_daqreader_mfdaq_cedspike2_obj.cedspike2filelist2smrfile(epochfiles);
                        fid = fopen(filename,'r');
                        p = [];
                        for i=1:numel(channel),
                            Info=SONChannelInfo(fid,channel(i));
    						datatype = 'int16';
	    					datasize = 16;
                            s2 = Info.scale/6553.6;
                            o2 = Info.offset;
				    		p = [p ; [-o2/s2 s2]];
                        end;
                        fclose(fid);
					case {'auxiliary_in'}
						datatype = 'uint16';
						datasize = 16;
                        p = repmat([0 1],numel(channel),1);
					case {'time'},
						datatype = 'float64';
						datasize = 64;
                        p = repmat([0 1],numel(channel),1);
                    case {'digital_in','digital_out'},
						datatype = 'char';
						datasize = 8;
                        p = repmat([0 1],numel(channel),1);
					case {'eventmarktext','event','marker','text'}, 
						datatype = 'float64';
						datasize = 64;
                        p = repmat([0 1],numel(channel),1);
					otherwise,
						error(['Unknown channel type ' channeltype '.']);
				end; %
		end;

	end % methods

	methods (Static)  % helper functions

		function smrfile = cedspike2filelist2smrfile(filelist)
			% CEDSPIKE2SPIKELIST2SMRFILE - Identify the .SMR file out of a file list
			% 
			% FILENAME = CEDSPIKE2FILELIST2SMRFILE(FILELIST)
			%
			% Given a cell array of strings FILELIST with full-path file names,
			% this function identifies the first file with an extension '.smr' (case insensitive)
			% and returns the result in FILENAME (full-path file name).
				for k=1:numel(filelist),
					[pathpart,filenamepart,extpart] = fileparts(filelist{k});
					if strcmpi(extpart,'.smr'),
						smrfile = filelist{k}; % assume only 1 file
						return;
					end; % got the .smr file
				end
				error(['Could not find any .smr file in the file list.']);
		end

		function channeltype = cedspike2headertype2mfdaqchanneltype(cedspike2channeltype)
		% CEDSPIKE2HEADERTYPE2MFDAQCHANNELTYPE- Convert between Intan headers and the ndi.daq.system.mfdaq channel types 
		%
		% CHANNELTYPE = CEDSPIKE2HEADERTYPE2MFDAQCHANNELTYPE(CEDSPIKE2CHANNELTYPE)
		% 
		% Given an Intan header file type, returns the standard ndi.daq.system.mfdaq channel type

			switch (cedspike2channeltype),
				case {1,9},
					% 1 is integer, 9 is single precision floating point
					channeltype = 'analog_in';
				case {2,3,4},
					channeltype = 'event'; % event indicator
						% 2 - positive-to-negative transition
						% 3 - negative-to-positive transition
						% 4 - either transition
				case {5,6},
						% 6 - wavemark, a Spike2-detected event
					channeltype = 'marker';
                case {8},
                    channeltype = 'text';
				case {7,9},
					error(['do not know this event yet--programmer should look it up.']);
				otherwise,
					error(['Could not convert channeltype ' cedspike2channeltype '.']);
			end;

		end % mfdaqchanneltype2cedspike2headertype()

	end % methods (Static)
end

File: ndi-python/ndi/daq/reader/mfdaq/cedspike2.py
"""
NDI DAQ Reader MFDAQ CED Spike2 - Device driver for CED Spike2.

This class reads data from CED Spike2 .SMR or .SON file formats.

It depends on sigTOOL by Malcolm Lidierth (http://sigtool.sourceforge.net).
"""

from typing import List, Tuple, Dict, Any, Optional, Union
import os
import numpy as np
from ..mfdaq import MFDAQReader


class CEDSpike2(MFDAQReader):
    """
    CED Spike2 reader for CED SMR/SON file format.

    This class reads data from CED Spike2 .SMR or .SON file formats,
    supporting analog, digital, event, marker, and text channels.

    Examples:
        >>> from ndi.daq.reader.mfdaq import CEDSpike2
        >>> reader = CEDSpike2()
        >>> channels = reader.getchannelsepoch(epochfiles)
    """

    def __init__(self, *args):
        """
        Create a new CED Spike2 reader.

        Args:
            *args: Arguments passed to MFDAQReader constructor
        """
        super().__init__(*args)

    def getchannelsepoch(self, epochfiles: List[str]) -> List[Dict[str, Any]]:
        """
        List the channels that are available on this CED Spike2 device.

        Args:
            epochfiles: List of file paths for this epoch

        Returns:
            List of channel dictionaries with fields:
            - 'name': The name of the channel (e.g., 'ai1', 'ev1')
            - 'type': The type of data stored in the channel
            - 'time_channel': The channel number that has time information
        """
        channels = []

        filename = self.cedspike2filelist2smrfile(epochfiles)
        header = self._read_ced_header(filename)

        if not header or 'channelinfo' not in header or not header['channelinfo']:
            return channels

        # Process each channel in header
        for ch_info in header['channelinfo']:
            # Convert CED type to MFDAQ type
            mfdaq_type = self.cedspike2headertype2mfdaqchanneltype(ch_info['kind'])

            # Get MFDAQ prefix
            from ...system.mfdaq import MFDAQSystem
            prefix = MFDAQSystem.mfdaq_prefix(mfdaq_type)

            channel_number = ch_info['number']

            # Add main channel
            channels.append({
                'name': f'{prefix}{channel_number}',
                'type': mfdaq_type,
                'time_channel': channel_number
            })

            # For analog channels, add a separate time channel
            if mfdaq_type == 'analog_in':
                channels.append({
                    'name': f't{channel_number}',
                    'type': 'time',
                    'time_channel': channel_number
                })

        return channels

    def verifyepochprobemap(self, epochprobemap: Any, epochfiles: List[str]) -> Tuple[bool, str]:
        """
        Verifies that an EPOCHPROBEMAP is compatible with a given device and the data on disk.

        Args:
            epochprobemap: NDI EpochProbeMap object
            epochfiles: List of file paths for this epoch

        Returns:
            Tuple of (is_valid, error_message)
        """
        # TODO: Implement verification when EpochProbeMap is fully defined
        return True, ''

    def readchannels_epochsamples(self, channeltype: Any, channel: List[int],
                                  epochfiles: List[str], s0: int, s1: int) -> np.ndarray:
        """
        Read the data based on specified channels.

        Args:
            channeltype: Type of channel to read (string or list of strings)
            channel: List of channel numbers to read (1-indexed)
            epochfiles: List of file paths for this epoch
            s0: Start sample (1-indexed)
            s1: End sample (1-indexed)

        Returns:
            Data array where each column contains data from an individual channel
        """
        # Ensure channeltype is a list
        if not isinstance(channeltype, list):
            channeltype = [channeltype] * len(channel)

        # Check that all channels are the same type
        unique_types = list(set(channeltype))
        if len(unique_types) != 1:
            raise ValueError('Only one type of channel may be read per function call at present.')

        filename = self.cedspike2filelist2smrfile(epochfiles)

        # Get sample rates
        sr = self.samplerate(epochfiles, channeltype, channel)
        sr_unique = list(set(sr))
        if len(sr_unique) != 1:
            raise ValueError('Do not know how to handle different sampling rates across channels.')

        sr_val = sr_unique[0]

        # Convert samples to time
        t0 = (s0 - 1) / sr_val
        t1 = (s1 - 1) / sr_val

        # Handle infinite times
        if np.isinf(t0) or np.isinf(t1):
            t0_orig = t0
            t1_orig = t1
            t0t1_here = self.t0_t1(epochfiles)

            if np.isinf(t0_orig):
                if t0_orig < 0:
                    t0 = t0t1_here[0][0]
                elif t0_orig > 0:
                    t0 = t0t1_here[0][1]

            if np.isinf(t1_orig):
                if t1_orig < 0:
                    t1 = t0t1_here[0][0]
                elif t1_orig > 0:
                    t1 = t0t1_here[0][1]

        # Read data for each channel
        data = []
        ct = channeltype[0]

        for ch in channel:
            if ct == 'time':
                # Read time data
                ch_data, _, _, _, time_data = self._read_ced_datafile(filename, ch, t0, t1)
                if data == []:
                    data = time_data.reshape(-1, 1)
                else:
                    data = np.concatenate([data, time_data.reshape(-1, 1)], axis=1)
            else:
                # Read regular data
                ch_data = self._read_ced_datafile(filename, ch, t0, t1)
                if data == []:
                    data = ch_data.reshape(-1, 1)
                else:
                    data = np.concatenate([data, ch_data.reshape(-1, 1)], axis=1)

        if data == []:
            data = np.array([])

        return data

    def readevents_epochsamples_native(self, channeltype: Any, channel: List[int],
                                      epochfiles: List[str], t0: float, t1: float) -> Tuple[Union[np.ndarray, List], Union[np.ndarray, List]]:
        """
        Read events or markers of specified channels for a specified epoch.

        Args:
            channeltype: Type of channel to read ('event', 'marker', 'text')
            channel: List of channel numbers to read
            epochfiles: List of file paths for this epoch
            t0: Start time
            t1: End time

        Returns:
            Tuple of (timestamps, data) where:
            - timestamps: Time of each event (or list of arrays for multiple channels)
            - data: Event data (1 for events, marker codes for markers, text for text)
        """
        # Ensure channeltype is a list
        if not isinstance(channeltype, list):
            channeltype = [channeltype] * len(channel)

        timestamps = []
        data = []

        filename = self.cedspike2filelist2smrfile(epochfiles)

        for i, ch in enumerate(channel):
            ct = channeltype[i]

            # Read data
            ch_data, _, _, _, ch_timestamps = self._read_ced_datafile(filename, ch, t0, t1)

            # Process based on channel type
            if ct == 'event':
                # Events are just timestamps with value 1
                ch_data = np.ones(len(ch_data))
            elif ct == 'marker':
                # Markers are 4-byte codes
                if len(ch_data) > 0:
                    # Convert 4 bytes to 32-bit integer
                    marker_values = np.sum(
                        ch_data * np.array([2**0, 2**8, 2**16, 2**24]).reshape(1, -1),
                        axis=1
                    )
                    ch_data = marker_values
            elif ct == 'text':
                # Text data - convert to list of strings
                if len(ch_data) > 0:
                    ch_data = [row.tobytes().decode('utf-8', errors='ignore').strip('\x00') for row in ch_data]

            timestamps.append(ch_timestamps)
            data.append(ch_data)

        # If single channel, return arrays directly instead of lists
        if len(channel) == 1:
            timestamps = timestamps[0]
            data = data[0]

        return timestamps, data

    def samplerate(self, epochfiles: List[str], channeltype: Any,
                  channel: List[int]) -> List[float]:
        """
        Get the sample rate for specific epoch and channel.

        Args:
            epochfiles: List of file paths for this epoch
            channeltype: Type of channel (string or list of strings)
            channel: List of channel numbers

        Returns:
            List of sample rates for each channel
        """
        filename = self.cedspike2filelist2smrfile(epochfiles)

        sr = []
        for ch in channel:
            sample_interval = self._read_ced_sampleinterval(filename, ch)
            sr.append(1.0 / sample_interval)

        return sr

    def t0_t1(self, epochfiles: List[str]) -> List[List[float]]:
        """
        Return the t0_t1 (beginning and end) epoch times for an epoch.

        Args:
            epochfiles: List of file paths for this epoch

        Returns:
            List containing [t0, t1] pair in seconds
        """
        filename = self.cedspike2filelist2smrfile(epochfiles)
        header = self._read_ced_header(filename)

        if not header or 'fileinfo' not in header:
            return [[0, 0]]

        # Calculate total time from header
        # t1 = dTimeBase * maxFTime * usPerTime
        fileinfo = header['fileinfo']
        t0 = 0.0  # Note: first sample is actually at 0 + 1/4 * sample_interval
        t1 = fileinfo['dTimeBase'] * fileinfo['maxFTime'] * fileinfo['usPerTime']

        return [[t0, t1]]

    def underlying_datatype(self, epochfiles: List[str], channeltype: str,
                           channel: List[int]) -> Tuple[str, Union[List[float], np.ndarray], int]:
        """
        Get the underlying data type for a channel in an epoch.

        Args:
            epochfiles: List of file paths for this epoch
            channeltype: Type of channel (string)
            channel: List of channel numbers

        Returns:
            Tuple of (datatype, polynomial, datasize) where:
            - datatype: Type suitable for fread/fwrite
            - polynomial: Conversion polynomial [offset, scale] (may be per-channel)
            - datasize: Sample size in bits
        """
        if channeltype in ['analog_in', 'analog_out']:
            # For analog channels, need to read channel-specific scaling
            filename = self.cedspike2filelist2smrfile(epochfiles)

            p = []
            for ch in channel:
                # Read channel info to get scale and offset
                info = self._read_ced_channelinfo(filename, ch)
                s2 = info['scale'] / 6553.6
                o2 = info['offset']
                p.append([-o2/s2, s2])

            datatype = 'int16'
            datasize = 16

            if len(p) == 1:
                p = p[0]
            else:
                p = np.array(p)

        elif channeltype == 'auxiliary_in':
            datatype = 'uint16'
            datasize = 16
            p = [0, 1]

        elif channeltype == 'time':
            datatype = 'float64'
            datasize = 64
            p = [0, 1]

        elif channeltype in ['digital_in', 'digital_out']:
            datatype = 'char'
            datasize = 8
            p = [0, 1]

        elif channeltype in ['eventmarktext', 'event', 'marker', 'text']:
            datatype = 'float64'
            datasize = 64
            p = [0, 1]

        else:
            raise ValueError(f'Unknown channel type {channeltype}.')

        return datatype, p, datasize

    # Static helper methods

    @staticmethod
    def cedspike2filelist2smrfile(filelist: List[str]) -> str:
        """
        Identify the .SMR file out of a file list.

        Args:
            filelist: List of full-path file names

        Returns:
            Path to the .smr file

        Raises:
            ValueError: If no .smr file found
        """
        for filepath in filelist:
            _, ext = os.path.splitext(filepath)
            if ext.lower() == '.smr':
                return filepath

        raise ValueError('Could not find any .smr file in the file list.')

    @staticmethod
    def cedspike2headertype2mfdaqchanneltype(cedspike2channeltype: int) -> str:
        """
        Convert between CED Spike2 headers and the ndi.daq.reader.mfdaq channel types.

        Args:
            cedspike2channeltype: CED channel type code (integer)

        Returns:
            MFDAQ channel type string
        """
        conversion = {
            1: 'analog_in',      # Integer waveform
            9: 'analog_in',      # Single precision floating point
            2: 'event',          # Positive-to-negative transition
            3: 'event',          # Negative-to-positive transition
            4: 'event',          # Either transition
            5: 'marker',         # Marker
            6: 'marker',         # Wavemark (Spike2-detected event)
            8: 'text'            # Text marker
        }

        if cedspike2channeltype == 7:
            raise ValueError('Channel type 7 not yet supported - programmer should look it up.')

        if cedspike2channeltype not in conversion:
            raise ValueError(f'Could not convert channeltype {cedspike2channeltype}.')

        return conversion[cedspike2channeltype]

    # Private helper methods for file I/O (placeholders for external dependencies)

    def _read_ced_header(self, filename: str) -> Dict[str, Any]:
        """
        Read CED SMR/SON header.

        Args:
            filename: Path to SMR/SON file

        Returns:
            Header dictionary

        Note:
            This is a placeholder for the external read_CED_SOMSMR_header function.
            TODO: Implement or import from sigTOOL when available.
        """
        # TODO: Implement actual CED header reading
        return {
            'channelinfo': [],
            'fileinfo': {
                'dTimeBase': 1.0e-6,
                'maxFTime': 1000000,
                'usPerTime': 1
            }
        }

    def _read_ced_datafile(self, filename: str, channel: int,
                          t0: float, t1: float) -> Union[np.ndarray, Tuple[np.ndarray, Any, Any, Any, np.ndarray]]:
        """
        Read CED SMR/SON data file.

        Args:
            filename: Path to SMR/SON file
            channel: Channel number
            t0: Start time
            t1: End time

        Returns:
            Data array, or tuple of (data, _, _, _, timestamps) for event channels

        Note:
            This is a placeholder for the external read_CED_SOMSMR_datafile function.
            TODO: Implement or import from sigTOOL when available.
        """
        # TODO: Implement actual CED data reading
        return np.array([])

    def _read_ced_sampleinterval(self, filename: str, channel: int) -> float:
        """
        Read CED SMR/SON sample interval.

        Args:
            filename: Path to SMR/SON file
            channel: Channel number

        Returns:
            Sample interval in seconds

        Note:
            This is a placeholder for the external read_CED_SOMSMR_sampleinterval function.
            TODO: Implement or import from sigTOOL when available.
        """
        # TODO: Implement actual sample interval reading
        return 0.001  # Default 1 ms

    def _read_ced_channelinfo(self, filename: str, channel: int) -> Dict[str, Any]:
        """
        Read CED SMR/SON channel info.

        Args:
            filename: Path to SMR/SON file
            channel: Channel number

        Returns:
            Channel info dictionary with 'scale' and 'offset'

        Note:
            This is a placeholder for the external SONChannelInfo function.
            TODO: Implement or import from sigTOOL when available.
        """
        # TODO: Implement actual channel info reading
        return {
            'scale': 1.0,
            'offset': 0.0
        }

    def __repr__(self) -> str:
        """String representation."""
        return f"CEDSpike2(name='{self.name if hasattr(self, 'name') else ''}')"

    def __str__(self) -> str:
        """String representation."""
        return self.__repr__()
